drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -771,11 +771,90 @@ static int dwt_filter(const struct device *dev,
 	return -ENOTSUP;
 }
 
-static int dwt_set_power(const struct device *dev, int16_t dbm)
+uint8_t dwt_encode_power_tx(int16_t power)
+{
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
+	uint8_t fineTuneVal = 0;
+	uint8_t result;
+
+	if (power < 0) {
+		coarse--;
+	} else if (power > DWT_TX_POWER_CTRL_MAX_MDB) {
+		coarse = 0;
+		fineTuneVal = 31;
+	} else {
+		coarse--;
+		while (power >= DWT_TX_POWER_CTRL_COARSE_STEPS_MDB && coarse > 0) {
+			power -= DWT_TX_POWER_CTRL_COARSE_STEPS_MDB;
+			coarse--;
+		}
+		fineTuneVal = (uint8_t)(power / DWT_TX_POWER_CTRL_FINE_STEPS_MDB) &
+			      DWT_TX_POWER_CTRL_FINE_MASK;
+	}
+
+	result = (coarse << 5) | fineTuneVal;
+
+	return result;
+}
+
+static int dwt_set_txpower(const struct device *dev, int16_t dbm)
 {
 	struct dwt_context *ctx = dev->data;
+	uint8_t chan = ctx->rf_cfg.channel;
+	uint32_t tx_power = 0;
+	int16_t mdbm = dbm * 10;
+
+	if (ctx->rf_cfg.dis_stxp) {
+		/* Manual Transmit Power Control */
+		uint16_t phr_sd_mdB;
+		uint8_t phr_sd;
+
+		if (ctx->rf_cfg.prf == DWT_PRF_64M) {
+			phr_sd_mdB = dwt_txpwr_stxp1_64_mdB[dwt_ch_to_cfg[chan] * 4];
```

Comment:
Yes, you are right, two-dimensional is way more intuitive.
But I cannot remove all of them since the I need to keep the values for the different channels since the channels can be changed at run time.